File: 36-financial-operations/payments-platform/state_machine.py
# Companion code for "The Backend of Luck" - Chapter 36, Financial Operations.
# https://thebackendofluck.com | https://github.com/thebackendofluck/book
# SPDX-License-Identifier: Apache-2.0
#
# FOR TESTING AND EVALUATION ONLY. NOT FOR PRODUCTION USE.
# Published to demonstrate the patterns explained in the book. This code is
# not certified for real-money gaming: operating a gambling platform requires
# your own licence, independent test-lab certification (GLI, eCOGRA or
# equivalent) and regulator approval.

"""
Payment state machine.

Models the full lifecycle of a payment:

    STARTED
      |
      +--[validation ok]---> PENDING
      |                         |
      |                   [PSP call initiated]
      |                         |
      |                      PROCESSING
      |                     /    |    \
      |            [3DS req]  [ok]  [error]
      |                |       |       |
      |             VERIFY  SUCCEEDED  FAILED
      |                |
      |          [auth complete]
      |                |
      |           PROCESSING --> SUCCEEDED / FAILED
      |
      +--[timeout]---> ABANDONED
      |
      +--[void request on PROCESSING] --> VOIDING --> VOIDED / VOID_FAILED

Refunds are modelled as a separate terminal state on SUCCEEDED payments.
"""

from __future__ import annotations

from typing import Optional, TypeVar

from models import Payment, PaymentStatus, WithdrawalStatus, Withdrawal

PaymentT = TypeVar("PaymentT", bound=Payment)
WithdrawalT = TypeVar("WithdrawalT", bound=Withdrawal)


# ---------------------------------------------------------------------------
# Allowed transitions
# ---------------------------------------------------------------------------

PAYMENT_TRANSITIONS: dict[PaymentStatus, set[PaymentStatus]] = {
    PaymentStatus.STARTED:     {PaymentStatus.PENDING, PaymentStatus.FAILED, PaymentStatus.ABANDONED},
    PaymentStatus.PENDING:     {PaymentStatus.PROCESSING, PaymentStatus.FAILED, PaymentStatus.ABANDONED},
    PaymentStatus.PROCESSING:  {PaymentStatus.SUCCEEDED, PaymentStatus.FAILED, PaymentStatus.VERIFY, PaymentStatus.VOIDING, PaymentStatus.ABANDONED},
    PaymentStatus.VERIFY:      {PaymentStatus.PROCESSING, PaymentStatus.FAILED, PaymentStatus.ABANDONED},
    PaymentStatus.VOIDING:     {PaymentStatus.VOIDED, PaymentStatus.VOID_FAILED},
    # Terminal states — no outbound transitions except SUCCEEDED -> REFUNDED
    PaymentStatus.SUCCEEDED:   {PaymentStatus.REFUNDED},
    PaymentStatus.FAILED:      set(),
    PaymentStatus.ABANDONED:   {PaymentStatus.SUCCEEDED},   # late PSP notification
    PaymentStatus.VOIDED:      set(),
    PaymentStatus.VOID_FAILED: set(),
    PaymentStatus.REFUNDED:    set(),
}

WITHDRAWAL_TRANSITIONS: dict[WithdrawalStatus, set[WithdrawalStatus]] = {
    WithdrawalStatus.STARTED:      {WithdrawalStatus.PENDING, WithdrawalStatus.INVALIDATED},
    WithdrawalStatus.PENDING:      {WithdrawalStatus.REVIEW, WithdrawalStatus.ACCEPTED, WithdrawalStatus.REJECTED},
    WithdrawalStatus.REVIEW:       {WithdrawalStatus.ACCEPTED, WithdrawalStatus.REJECTED, WithdrawalStatus.TIMED_OUT},
    WithdrawalStatus.ACCEPTED:     {WithdrawalStatus.PROCESSING},
    WithdrawalStatus.PROCESSING:   {WithdrawalStatus.REVERSED, WithdrawalStatus.FAILED},
    WithdrawalStatus.FAILED:       {WithdrawalStatus.PENDING},   # retry path
    WithdrawalStatus.REVERSED:     set(),
    WithdrawalStatus.REJECTED:     set(),
    WithdrawalStatus.TIMED_OUT:    set(),
    WithdrawalStatus.INVALIDATED:  set(),
}


# ---------------------------------------------------------------------------
# Exceptions
# ---------------------------------------------------------------------------


class InvalidTransitionError(Exception):
    def __init__(self, current: PaymentStatus, target: PaymentStatus) -> None:
        super().__init__(
            f"Cannot transition payment from {current.value} to {target.value}"
        )


class InvalidWithdrawalTransitionError(Exception):
    def __init__(self, current: WithdrawalStatus, target: WithdrawalStatus) -> None:
        super().__init__(
            f"Cannot transition withdrawal from {current.value} to {target.value}"
        )


# ---------------------------------------------------------------------------
# State machine helpers
# ---------------------------------------------------------------------------


class PaymentStateMachine:
    """Validates and applies state transitions for Payment objects."""

    def transition(
        self,
        payment: PaymentT,
        new_status: PaymentStatus,
        failure_type: Optional[str] = None,
        failure_reason: Optional[str] = None,
    ) -> PaymentT:
        allowed = PAYMENT_TRANSITIONS.get(payment.status, set())
        if new_status not in allowed:
            raise InvalidTransitionError(payment.status, new_status)

        from datetime import datetime, timezone
        from models import FailureInfo

        updated = payment.model_copy(
            update={
                "status": new_status,
                "updated_at": datetime.now(timezone.utc),
                "failure_info": FailureInfo(
                    failure_type=failure_type,
                    failure_reason=failure_reason,
                )
                if failure_type
                else payment.failure_info,
            }
        )
        return updated

    def can_transition(self, payment: Payment, new_status: PaymentStatus) -> bool:
        return new_status in PAYMENT_TRANSITIONS.get(payment.status, set())

    def pending(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.PENDING)

    def processing(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.PROCESSING)

    def verify(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.VERIFY)

    def succeed(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.SUCCEEDED)

    def fail(self, payment: PaymentT, failure_type: str, reason: str) -> PaymentT:
        return self.transition(payment, PaymentStatus.FAILED, failure_type, reason)

    def abandon(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.ABANDONED)

    def void(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.VOIDING)

    def voided(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.VOIDED)

    def refund(self, payment: PaymentT) -> PaymentT:
        return self.transition(payment, PaymentStatus.REFUNDED)


class WithdrawalStateMachine:
    """Validates and applies state transitions for Withdrawal objects."""

    def transition(
        self,
        withdrawal: WithdrawalT,
        new_status: WithdrawalStatus,
        error_message: Optional[str] = None,
    ) -> WithdrawalT:
        allowed = WITHDRAWAL_TRANSITIONS.get(withdrawal.status, set())
        if new_status not in allowed:
            raise InvalidWithdrawalTransitionError(withdrawal.status, new_status)

        from datetime import datetime, timezone

        updated = withdrawal.model_copy(
            update={
                "status": new_status,
                "updated_at": datetime.now(timezone.utc),
                **({"error_message": error_message} if error_message else {}),
            }
        )
        return updated

    def can_transition(self, withdrawal: Withdrawal, new_status: WithdrawalStatus) -> bool:
        return new_status in WITHDRAWAL_TRANSITIONS.get(withdrawal.status, set())

    def submit(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.PENDING)

    def flag_review(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.REVIEW)

    def accept(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.ACCEPTED)

    def reject(self, withdrawal: WithdrawalT, reason: str) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.REJECTED, reason)

    def process(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.PROCESSING)

    def reverse(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.REVERSED)

    def fail(self, withdrawal: WithdrawalT, reason: str) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.FAILED, reason)

    def invalidate(self, withdrawal: WithdrawalT) -> WithdrawalT:
        return self.transition(withdrawal, WithdrawalStatus.INVALIDATED)

File: 10-complete-platform-architecture/platform-core/workers/payments-worker/src/state-machine.ts
// Companion code for "The Backend of Luck" - Chapter 10, Complete Platform Architecture.
// https://thebackendofluck.com | https://github.com/thebackendofluck/book
// SPDX-License-Identifier: Apache-2.0
//
// FOR TESTING AND EVALUATION ONLY. NOT FOR PRODUCTION USE.
// Published to demonstrate the patterns explained in the book. This code is
// not certified for real-money gaming: operating a gambling platform requires
// your own licence, independent test-lab certification (GLI, eCOGRA or
// equivalent) and regulator approval.

/**
 * state-machine.ts
 * ----------------
 * Payment state transitions for iGaming deposit and withdrawal flows.
 *
 * State model:
 *   PENDING → PROCESSING → COMPLETED | FAILED | CANCELLED
 *   COMPLETED → REFUNDED (for chargebacks)
 *
 * All transitions are validated before application; invalid transitions
 * return an error result so the caller can handle idempotent retries.
 */

export type PaymentStatus =
  | "PENDING"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED"
  | "CANCELLED"
  | "REFUNDED";

export type PaymentType = "DEPOSIT" | "WITHDRAWAL";

export interface Payment {
  id: string;
  playerId: string;
  type: PaymentType;
  status: PaymentStatus;
  amount: number;       // minor units (e.g. pence / cents)
  currency: string;     // ISO 4217
  pspRef?: string;
  pspName?: string;
  failureReason?: string;
  createdAt: number;
  updatedAt: number;
  metadata?: Record<string, string>;
}

// ---------------------------------------------------------------------------
// Allowed transitions
// ---------------------------------------------------------------------------

const ALLOWED_TRANSITIONS: Record<PaymentStatus, PaymentStatus[]> = {
  PENDING:    ["PROCESSING", "CANCELLED", "FAILED"],
  PROCESSING: ["COMPLETED", "FAILED", "CANCELLED"],
  COMPLETED:  ["REFUNDED"],
  FAILED:     [],
  CANCELLED:  [],
  REFUNDED:   [],
};

export type TransitionResult =
  | { ok: true; payment: Payment }
  | { ok: false; error: string };

/**
 * Attempt a status transition.  Returns the updated payment on success
 * or an error description on invalid transitions.
 */
export function transition(
  payment: Payment,
  toStatus: PaymentStatus,
  meta?: { pspRef?: string; failureReason?: string },
): TransitionResult {
  const allowed = ALLOWED_TRANSITIONS[payment.status];

  if (!allowed.includes(toStatus)) {
    return {
      ok: false,
      error: `Invalid transition: ${payment.status} → ${toStatus} for payment ${payment.id}`,
    };
  }

  const updated: Payment = {
    ...payment,
    status: toStatus,
    updatedAt: Date.now(),
    ...(meta?.pspRef ? { pspRef: meta.pspRef } : {}),
    ...(meta?.failureReason ? { failureReason: meta.failureReason } : {}),
  };

  return { ok: true, payment: updated };
}

/**
 * Create a new payment in PENDING state.
 */
export function createPayment(params: {
  id: string;
  playerId: string;
  type: PaymentType;
  amount: number;
  currency: string;
  metadata?: Record<string, string>;
}): Payment {
  const now = Date.now();
  return {
    ...params,
    status: "PENDING",
    createdAt: now,
    updatedAt: now,
  };
}

/**
 * Determines whether a payment is in a terminal state.
 */
export function isTerminal(payment: Payment): boolean {
  return ["COMPLETED", "FAILED", "CANCELLED", "REFUNDED"].includes(payment.status);
}
